crates/stackable-webhook/src/servers/conversion_webhook.rs

```diff
@@ -0,0 +1,185 @@
+pub use kube::core::conversion::ConversionReview;
+use kube::{
+    Api, Client, ResourceExt,
+    api::{Patch, PatchParams},
+};
+use snafu::{ResultExt, Snafu, ensure};
+use tokio::sync::oneshot;
+use x509_cert::Certificate;
+
+use super::{Webhook, WebhookError};
+use crate::WebhookServerOptions;
+
+#[derive(Debug, Snafu)]
+pub enum ConversionWebhookError {
+    #[snafu(display("failed to send initial CRD reconcile heartbeat"))]
+    SendInitialReconcileHeartbeat,
+
+    #[snafu(display("failed to patch CRD {crd_name:?}"))]
+    PatchCrd {
+        source: kube::Error,
+        crd_name: String,
+    },
+}
+
+pub struct ConversionWebhook<H> {
+    crds_and_handlers: Vec<(CustomResourceDefinition, H)>,
+    disable_crd_maintenance: bool,
+    client: Client,
+    /// The field manager used when maintaining the CRDs.
+    field_manager: String,
+    // This channel can only be used exactly once. The sender's send method consumes self, and
+    // as such, the sender is wrapped in an Option to be able to call take to consume the inner
+    // value.
+    initial_reconcile_tx: Option<oneshot::Sender<()>>,
+}
+
+impl<H> ConversionWebhook<H> {
+    pub fn new(
+        crds_and_handlers: impl IntoIterator<Item = (CustomResourceDefinition, H)>,
+        disable_crd_maintenance: bool,
+        client: Client,
+        field_manager: String,
+    ) -> (Self, oneshot::Receiver<()>) {
+        let (initial_reconcile_tx, initial_reconcile_rx) = oneshot::channel();
+
+        (
+            Self {
+                crds_and_handlers: crds_and_handlers.into_iter().collect(),
+                disable_crd_maintenance,
+                client,
+                field_manager,
+                initial_reconcile_tx: Some(initial_reconcile_tx),
+            },
+            initial_reconcile_rx,
+        )
+    }
+}
+
+#[async_trait]
+impl<H> Webhook for ConversionWebhook<H>
+where
+    H: FnOnce(ConversionReview) -> ConversionReview + Clone + Send + Sync + 'static,
+{
+    fn register_routes(&self, mut router: Router) -> Router {
+        for (crd, handler) in self.crds_and_handlers.clone() {
+            let crd_name = crd.name_any();
+            let handler_fn = |Json(review): Json<ConversionReview>| async {
+                let review = handler(review);
+                Json(review)
+            };
+
+            let route = format!("/convert/{crd_name}");
+            router = router.route(&route, post(handler_fn));
+        }
+
+        router
+    }
+
+    async fn handle_certificate_rotation(
+        &mut self,
+        _new_certificate: &Certificate,
+        new_ca_bundle: &ByteString,
+        options: &WebhookServerOptions,
+    ) -> Result<(), WebhookError> {
+        if self.disable_crd_maintenance {
+            return Ok(());
+        }
+
+        tracing::info!(
+            k8s.crd.names = ?self.crds_and_handlers.iter().map(|(crd, _)| crd.name_any()).collect::<Vec<_>>(),
+            "reconciling custom resource definitions"
+        );
```

Comment:
The same attribute can be sent with multiple values (though maybe that depends on the subscriber?).

So this could be inverted to iterating with forEach, and emitting the event.

Something like this:

```rust
self.crds_and_handlers.forEach(|(crd, _)| {
    tracing::info!(
        k8s.crd.name = ?crd.name_any(),
        "reconciling custom resource definition"
    );
});
```

But further below, we iterate, and can just emit the even there at the time that it happens?

Reply:
This was existing code copied across, we both agreed to leave it as is for now